### DISCUSSION TITLE
Raspberry Pi 3B+?

### DISCUSSION THREAD

**Discussion by mcguirepr89** [mcguirepr89/BirdNET-Pi#48] [Q&A]
If any testers out there have access to a Raspberry Pi 3B+, it would be helpful in trying to plan for lower-power consumption to know it is possible. I assume the NoMachine won't work, but if the other things work, fixing NoMachine won't be too difficult.

Thanks in advance to any takers on this request --
My best,
Patrick

**Comment by CaiusX** [comment]
Hi @mcguirepr89 

Happy to install on one of my old Pi's hanging around as long as it won't interrupt my main system :).

Best
Dennis

Update>>>>>>>>>>>>

There's an Pi3 on my network with nomachine running & an old BirdNET-Lite installed - you should be able to get in with same details as the Pi4 - I can do a full clean installation of the latest BirdNET-Pi if you want, alternatively, you can take it from here.

best

**Reply by mcguirepr89** [reply]
Thank you, sir!! I'll take a look and let you know what I find --

**Reply by mcguirepr89** [reply]
Hello, my friend -- If the RPi3B+ is still available for testing, I wanted to ask if you would be willing to test out the pre-installed image on it? 

My best regards,
Patrick

**Reply by CaiusX** [reply]
Hi @mcguirepr89 

Spent some time trying to get it setup on a 3B+ but failing at network connection stage (so I can give you access).  Generally everything is very slow.....

Will persevere again in the morning.

Best

**Reply by mcguirepr89** [reply]
Thank you!! I appreciate your help

**Reply by CaiusX** [reply]
Hi @mcguirepr89 

Trying a fresh installation of 0.7 on a Rpi4 I'm able to get it up and running fine.

Going through the "Configure-Network" procedure as indicated doesn't give positive LAN wireless access (or any wireless access).  The wireless icon is the 
X|
|X

& hover over message is "Connection to dhcpcd lost".

Right clicking and selecting "Wireless and Wired Network Settings" goes nowhere.

The unit IS however on the network:

![image](https://user-images.githubusercontent.com/25373276/140702422-ab38d916-330a-49f5-ad26-7e06aa5748e6.png)


Please tell me what files I should send you to have a look at this?

I will now transfer this card to a Pi3b and see if I can get it accessible - stay tuned!

Best
AP/DC

**Reply by CaiusX** [reply]
Hi @mcguirepr89 

Transferring the SD to the Pi-3B+ -

Boots OK but super slow, once up and running nothing useful is accessible, Chrome web pages and BirdNET-Pi just hang.

Stopping the core services (stop_core_services.sh) from command line gets a usable Pi with ability to get  the DB view on Chrome, but also really slow, lots of no-response from mouse movement etc. 

I suspect Nomachine is quite a resource hog, so restarting without Nomachine

- Get Database View on Chrome with usable mouse
- Get Spectrogram view, but not "current" - about an hour old.
- BirdNET-Lite Log runs, seems to take about 47 seconds to process a 15 second sound sample wav file when nothing else going on, up to 2 minutes when mousing around etc.  
- With this timing we get a buildup of unprocessed audio files  in /BirdSongs/M-Y/Date (156 soundfiles and counting after <1 hour) as the BirdNET-Lite takes up most of the processing time.

Generally I think that the speed is being limited by the memory (only 1Gb RAM) and swapfiles taking most of the time.  
Loading the Tensorflow model for each analysis is probably also a rate limiting step.  

I had some success previously with trying to implement 

https://github.com/kahst/BirdNET-Lite/issues/5#issuecomment-907898741

> I was able to make Live Audio work as follows. Perhaps it is a naive approach but it seems robust.
> 
> * Create a script that records audio in 3 second clips and stashes them in a working directory. Only record up to a maximum number of clips eg. 2 files.
> * Modify the analysis script to scan for files in the  working directory - process the oldest file and then delete it.
> * Run these scripts as separate processes. The recording script will fill the directory and the analyses script will empty it.
> 
> Analysing a 3s clip on a Raspberry Pi 4 only takes about 0.5s, so this scheme should never fall behind real-time (ignoring the 3s delay for sampling the audio)

The key I think is "run these scripts as separate processes" - the BirdNET-Lite python script then sits in a loop waiting for the samples to appear, and then does the analysis/reporting as its own process, without interrupting/delaying any of the other services or having to load the Tensorflow model each run.  I never managed to get my head around this parallelism as some Linux genius diverted my attention to a different system.

If you think this RPI-3B route is worth pursuing (or the new Pi-Zero) maybe see if @CoreProduction is keen to help out on the Python side?  We need an Australian installation anyway to broaden the latitudinal and longitudinal user base and get to hear some Kookaburras....

I'm happy to keep this 3B+ installation live, will restart Nomachine & send you the login details by e-mail if you want to try have a look, but I'm not too hopeful that the system will cope ....

Best

**Reply by mcguirepr89** [reply]
Dennis! You are a tremendous help --- thank you so much

### RPi4
> Going through the "Configure-Network" procedure as indicated doesn't give positive LAN wireless access (or any wireless access). The wireless icon is the
X|
|X
& hover over message is "Connection to dhcpcd lost".
Right clicking and selecting "Wireless and Wired Network Settings" goes nowhere.
The unit IS however on the network:

This is because the desktop icon connects to the "dhcpcd" service, which is no longer present, but the LAN will be configured by 'systemd-networkd' for which there is no desktop panel tool to replace the "Wireless & Wired Network Settings" tool. 
It was probably stupid of me to keep that icon visible, but since it is possible to totally revert the networking to the standard dhcpcd, I didn't want to remove it completely.

### RPi3
> Generally I think that the speed is being limited by the memory (only 1Gb RAM) and swapfiles taking most of the time.
Loading the Tensorflow model for each analysis is probably also a rate limiting step.

I think you are exactly right. One thing you can try is enabling the zRAM swapping for the RPi3, since it desperately needs it.
Run the stop_core_services.sh and then [follow these screen shots ](https://github.com/mcguirepr89/BirdNET-Pi/issues/66#issuecomment-962503315 )and setup at least 2G for the zRAM swap and reboot to see if you get a more responsive system.

### Live Audio
I definitely don't understand this, since the arrangement that is suggested sounds a lot like how BirdNET-Pi operates? I'll need to reach out to @CoreProduction to see what exactly needs to be changed. My suspicion is that this approach is identical to adjusting the birdnet_recording.service script down to 3 second chunks (which would break the extraction.service, and indeed remove its purpose entirely). 

The other possibility may be to supplant BirdNET-Lite for the BirdNET-Demo InnoTruck model which is already built to handle streams -- perhaps it could be used to trigger the extraction.service on a longer recording? Anywho, I'll think on this, but adding adequate swap space to the RPi3 is the first step in seeing if pursuing less resource demanding services will be worth while.

I tried connecting to the machine via NoMachine, but without success.
THANK YOU THANK YOU THANK YOU!!!
:swan:

**Reply by CaiusX** [reply]
Hi @mcguirepr89 

Had "loadshedding" for the last couple of hours, but back now.

zRAM enabled and seems to be a lot better!!

You should be able to Nomachine either on the address 169.0.167.8 22684, same PW, or through the other RPi, local address 192.168.3.86.

Best

**Reply by mcguirepr89** [reply]
I've been able to connect and this seems promising. 

IF there were any way that you could add some cooling --the average temperature is around 85 degrees Celsius, so the CPU is throttling to prevent the machine from melting!-- and perhaps a slightly longer recording time (I've set it to 40 seconds with a 10 second extraction to test with for now) and disabling the livestream/icecast2 services altogether (may or may not be necessary), then I think this is working!

**Reply by mcguirepr89** [reply]
Looks like overclocking that high right off the bat prevented the thing from booting.  If you put the sd card into an sd card reader and pop it into the other BirdNET-Pi (Live from ZA :wink: ) then I can recover it and try something else?

**Reply by mcguirepr89** [reply]
Ok that's recovered -- thank you!

**Comment by ghost** [comment]
I think about to build a mobile device, perhaps with LCD display. I would be interested if the whole system runs with the PiZero. https://www.raspberrypi.com/products/raspberry-pi-zero/ or the PiPico https://www.raspberrypi.com/products/raspberry-pi-pico/ .

**Reply by mcguirepr89** [reply]
It will all depend on the pre-built TFLite binaries, I think. I'll look into it and see what I can find out

**Reply by ghost** [reply]
I think about a handheld device. Can also do a hardware design for that. Google supports also very basic 32-bis OS'es. But i am not sure to do all this, it is a lot of work and only profitable if the device is sold after the hardware design development. Perhaps so kickstarter found can be done. But I am only thinking about that because it is technically possible and one have to do this. I am sure that someone will build the first handheld device.

**Comment by CaiusX** [comment]
Done, shout when you finished.

**Reply by mcguirepr89** [reply]
Thanks! I unmounted the SD so you can remove it and it ought to boot properly now

**Comment by CaiusX** [comment]
Back n booted!

**Reply by mcguirepr89** [reply]
Ok I tried once more with a lower clock speed and smaller over_voltage setting -- we'll see if it boots

**Reply by mcguirepr89** [reply]
I killed it again :(
I can recover it and will think on this a bit more. Thank you so much for your help today!

**Comment by CaiusX** [comment]
Back in the Pi4

**Reply by mcguirepr89** [reply]
It doesn't appear to have booted in the RPi4, either. You'll need to mount that SD card onto another maching, open /boot/config.txt and remove the lines for 'overclock' and 'overvoltage' (or plug it into the BirdNET-Pi Live ZA where I will be happy to clean up my mess)

**Reply by mcguirepr89** [reply]
Thank you! All set -- to put back into the rpi4 now

**Comment by CaiusX** [comment]
This time properly. 

D
>

**Comment by Molokai** [comment]
I've been running successfully on rPi 3B+ for about a week, the last 4 days on the forms branch. It seems to work just fine. Currently using RECORDING_LENGTH=39 (I read somewhere here that I should be using multiples of 3).

It feels quite a bit more sluggish than running on the v4, but it seems to work just fine.

Let me know if there is anything else to do/report on for rpi 3b+ support.

**Reply by mcguirepr89** [reply]
Awesome!!! Thank you so much for the update! I'm hopeful that I can pul `forms` into main soon and then start a new testing branch for some of the feature requests that the `forms` branch has sparked.

**Comment by mcguirepr89** [comment]
Hello, all!!!

I wanted to post on this discussion to ask if anyone running BirdNET-Pi on the 3B+ notices issues when updating their settings in the web GUI via "Tools" > "Settings"? If so, could you please relay here what you typically observe?

Many thanks,
Patrick

**Reply by matthew73210** [reply]
Hey, running on a 3B+ here. Basiclly the interface hangs for a long time when you issue an update, what I do is a sudo reboot after about a min and it comes right back.
